jni/src/faiss_index_service.cpp

```diff
@@ -137,28 +218,35 @@ void BinaryIndexService::createIndex(
         throw std::runtime_error("Number of IDs does not match number of vectors");
     }
 
-    std::unique_ptr<faiss::IndexBinary> indexWriter(faissMethods->indexBinaryFactory(dim, indexDescription.c_str()));
-
     // Set thread count if it is passed in as a parameter. Setting this variable will only impact the current thread
     if(threadCount != 0) {
         omp_set_num_threads(threadCount);
     }
 
-    // Add extra parameters that cant be configured with the index factory
-    SetExtraParameters<faiss::IndexBinary, faiss::IndexBinaryIVF, faiss::IndexBinaryHNSW>(jniUtil, env, parameters, indexWriter.get());
-
-    // Check that the index does not need to be trained
-    if(!indexWriter->is_trained) {
-        throw std::runtime_error("Index is not trained");
-    }
+    faiss::IndexBinaryIDMap * idMap = reinterpret_cast<faiss::IndexBinaryIDMap *> (idMapAddress);
 
     // Add vectors
-    std::unique_ptr<faiss::IndexBinaryIDMap> idMap(faissMethods->indexBinaryIdMap(indexWriter.get()));
     idMap->add_with_ids(numVectors, inputVectors->data(), ids.data());
+}
+
+void BinaryIndexService::writeIndex(
+        std::string indexPath,
+        jlong idMapAddress
+    ) {
+
+    std::unique_ptr<faiss::IndexBinaryIDMap> idMap (reinterpret_cast<faiss::IndexBinaryIDMap *> (idMapAddress));
```

Comment:
Are we sure we want to transfer ownership of the index here? Seems this could be misleading

Reply:
The concern is more about if a code path is accessing it after write is complete. There is no mechanism currently to enforce it and this makes memory management simpler so keeping it as it is